### DISCUSSION TITLE
TVB GSOC 2025/2026

### DISCUSSION THREAD

**Discussion by liadomide** [the-virtual-brain/tvb-root#736] [Ideas]
## The Virtual Brain: An open-source simulator for whole brain network modeling.

Simulating the human brain is the holy grail of neuroscience – offering a pioneering tool for understanding how our brain works and how to deal with its disorders like stroke, epilepsy or neurodegenerative diseases like Alzheimer’s or Parkinson’s.

While large-scale research initiatives simulate neurons and small brain regions at the cellular level on massively parallel hardware, they are still years away from clinical applications.

The Virtual Brain ([TVB](https://www.thevirtualbrain.org/)) takes a different approach and reduces complexity on the micro level to attain the macro organization: a TVB model of a patient's brain generates sufficiently accurate EEG, MEG, BOLD and SEEG signals by reducing the complexity millionfold through methods from statistical physics. The key is TVB’s hybrid approach of merging individual anatomy from brain imaging data with state-of-the-art mathematical modeling.

### [1] New graphical widget(s) for JupyterLab

**Description:** 
In the TVB (https://www.thevirtualbrain.org/) ecosystem there is a new code repository called [tvb-widgets](https://github.com/the-virtual-brain/tvb-widgets) offering neat UI widgets for Jupyterlab environments. These widgets are compatible with TVB data formats and able to display these data in different forms: 2D, 3D. 
The purpose of this project is to implement a new set of widgets, which would allow users to work  in Jupyter Lab in 3D, with the connectivity matrices (nodes & edges), surfaces (cortical, subcortical, face), sensors and all data involved in a TVB simulation. 
Necessary features for this widget: select from a drive the list of available datesets, load/unload a dataset, display connectivity matrix, 3D surface, connections, colours, animated timeseries, etc. Of course, these new widgets have to run in a Jupyterlab notebook as well. We have most of these already in the existent widgets, but we are looking into maybe new libraries to use, better performance, better usability - so a renovation of some of the existent features.
Finally, it would be great to have all the widgets linked into the [tvb-ext-xircuits](https://github.com/the-virtual-brain/tvb-ext-xircuits) repository which is a Jupyterlab extension based on React JS. At the moment, only the [PhasePlaneWidget](https://drive.google.com/drive/folders/1DTMBSxbIcNdBBkSGCIEoyGqg3EIT4XDz) is linked there, but the rest could be added in a similar manner.

Examples of TVB data formats can be found on [Zenodo](https://zenodo.org/records/10128131). 
Check out our Jupyter [notebooks](https://github.com/the-virtual-brain/tvb-widgets/tree/main/notebooks) to play with the widgets we have available so far.

**Expected results:** A set of classes , with at least one demo Jupyter notebook, and unit tests.

**Preferred Tech keywords:** Python, IPywidgets, React JS, Jupyterlab, Jupyterlab extensions

**Skills level**: junior+, mid

**Mentors:** Lia Domide (lead), Paula Prodan, Teodora Misan (backup)

### [2] New components in tvb-ext-xircuits

**Description:** 
In the TVB (https://www.thevirtualbrain.org/) ecosystem there is a code repository called [tvb-ext-xircuits](https://github.com/the-virtual-brain/tvb-ext-xircuits) offering a graphical way to define, configure and run simulation workflows. 

The purpose of this project is to implement a new set of  xircuits components, in particular for running Parameter Space Explorations (so a range of simulations with little variation in 2-3 parameters).

**Expected results:** A set of new xircuits components , with at least one demo xircuits file.

**Preferred Tech keywords:** Python, xircuits, Jupyterlab4, Jupyterlab extensions

**Skills level**: junior+, mid

**Mentors:** Lia Domide (lead), Paula Prodan, Teodora Misan(backup)

**_All projects bellow can be tailored for different length, as the features/pages can be build incrementally._**

**Comment by deeptive** [comment]
Hello @liadomide 
This is Deepthi , I've found this project interesting and I am eagerly looking forward to contribute. I've already installed project on my system. It would be helpful if you could give me right approach to get started. Also , what exactly are you looking for in contributor?
Thank you

**Reply by liadomide** [reply]
Hi @deeptive 

We are happy to find of your interest!

To get started, please try to run the components that we linked above.

In a contributor we look for: somebody eager to  embrace the open-source world values, someone that enjoys clean & efficient code ;-)

**Reply by deeptive** [reply]
Hello @liadomide 
As suggested, I’ve spent time exploring the widgets. I've got an understanding of the codebase for the work developed so far.
I would be eager to contribute by addressing any existing bugs or implementing improvements. If possible, could you please assign me any issues to work on? 
Alternatively, if you prefer, I can identify relevant tasks myself.
 In that case, I would appreciate any suggestions you may have regarding areas where improvements could be made.
Thank you!.

**Reply by liadomide** [reply]
Thank you for your work!

Regarding widgets, we do not have a list of issue public, but these things would be useful from our perspective:
- update widgets to be compatible with the latest versions of the 3rd party dependencies (such as matplotlib)
- try to replace/get did of heavy dependencies (such as VTK)
- unify few widgets into a “Head Widget”, or a “simulation cockpit widget”

Does these help ?
Alternatively, you suggesting improvements, as a user touching widgets for the first time, is also more than welcome, as surely you have a fresh perspective.

**Reply by deeptive** [reply]
Thank you for the issues; they really help me get started!  

I was also thinking of making some enhancements to the UI. For example, in the *ConnectivityMatrix* widget, users currently have to select a cell in the matrix, scroll up, and enter a value in a separate input box. Instead, we could improve the user experience by making the matrix interactive, allowing users to enter values directly near the cell they clicked and save them in place. Enhancements like these would improve the overall experience and ensure a cleaner interface.  

I will start working on the issues you mentioned and, if possible, discuss and implement UI enhancements with you later.  

Thank you!

**Comment by deeptive** [comment]
Hello
@liadomide 
I've been working on the first issue you mentioned and identified that the following dependencies are outdated:
NumPy
ipywidgets
Matplotlib (Current: 3.10.0, Latest: 3.10.1)
Plotly

I have upgraded Matplotlib to version 3.10.1 and conducted testing. Since the previous version wasn't significantly outdated, everything continues to function as expected. I will now proceed with updating the other dependencies.

Please let me know if you have any specific concerns or guidelines for these updates.

**Reply by liadomide** [reply]
This update seems promising. 
Thanks for the news!
Please proceed, as you propose.

**Comment by deeptive** [comment]
Hi @liadomide @915-Misan-Teodora 
I previously contributed to this project.
Happy to see the project listed again this year and exicted to continue my work.
Made sure everything is compliant with program gudielines from my side this time.
Please let me know if you're okay with me contributing this year.
Thank you

**Reply by 915-Misan-Teodora** [reply]
Hello @deeptive,
Thanks for reaching out, great to hear you're interested in contributing again this year.
Yes, please feel free to contribute to our projects and thanks for checking the program guidelines.

**Reply by deeptive** [reply]
Hello @liadomide and @915-Misan-Teodora,
I noticed that this project idea appears to be the same as last year. Since I already have some familiarity with the existing codebase, I would be very interested in contributing further.
If possible, could you please suggest some tasks I can start working on? These could be existing issues suitable for a pull request or areas that would be helpful to explore in preparation for the final proposal. I would be happy to take them up and contribute accordingly.
Thank you, and I look forward to your guidance.
